### PR TITLE
Allow `epi_slide` to access `ref_time_value`

### PR DESCRIPTION
**Pull request by nmdefries**
This implements [Ryan's suggestion](https://delphi-org.slack.com/archives/C028AUR6MV1/p1659236179493529) for how to calculate `ref_time_value`, and makes sure `epi_slide` computations of all types can access `ref_time_value` and the group key using the same approaches as in https://github.com/cmu-delphi/epiprocess/pull/313 and https://github.com/cmu-delphi/epiprocess/pull/317.

I plan to remove the `n_mandatory_f_args` arg to `assert_sufficient_f_args` in another PR.

Closes https://github.com/cmu-delphi/epiprocess/issues/171.

### CODE REVIEW

**Comment by nmdefries** [comment]
There was some concern over the efficiency of finding `time_value`s and joining them onto the data, then sorting the data.

In the old implementation, we were already [doing a direct sort](https://github.com/cmu-delphi/epiprocess/blob/d0fc5b0e85aeab42e9ca4879849e3a49e9a80eaf/R/slide.R#L165). I've just moved that step to be a littler bit later in the new version. So that step hasn't changec computational complexity.

In the new code here, we do need to identify dates not already in the `epi_df`. I used a standard [`%in%` to identify these](https://github.com/cmu-delphi/epiprocess/blob/8bfb4d898d79ee27341b2ce8377eb94ee975e836/R/slide.R#L225) since we are doing the [same kind of search](https://github.com/cmu-delphi/epiprocess/blob/d0fc5b0e85aeab42e9ca4879849e3a49e9a80eaf/R/slide.R#L181) already.

The main addition, that is, something that could be slow that we're not already doing elsewhere in the `epi_slide` function`, are the handful of `bind_rows`, which will copy some data around. I don't know yet how this approach compares to doing a merge.

**Comment by nmdefries** [comment]
Doing some benchmarking, the slow steps of the new approach are dropping the `.real` column and filtering by the `.real` column.

**Comment by nmdefries** [comment]
When testing  that computations can access `ref_time_value`, I noticed that `slide` returns date columns in numeric format (`18672`, e.g.). It is [converted during `unlist`](https://github.com/cmu-delphi/epiprocess/blob/22b9b690a62ac10b551739c8a79f169969380eac/R/slide.R#L308). Not sure it's worth changing. We don't really expect users to return dates. They can also convert back to date format themselves.

**Comment by nmdefries** [comment]
The input data
```
set.seed(100)
n <- 5000L
x <- tibble::tribble(~geo_value, ~time_value, ~binary,
                     "x",       ceiling(runif(n, 0, 200)),   runif(n, 0, 100)^2) %>%
  tidyr::unnest(c(time_value,binary))

# Randomize order
x <- x[sample(nrow(x)), ] %>% as_epi_df()
```

Looking at a computation that doesn't require access to `ref_time_value` so that we can compare between the new and old versions (the new logic calculates `ref_time_value` even if it isn't used):
```
bench::mark(
  ref_value_calculated = {epi_slide(x, f = function(x, g, t) sum(x$binary),
                                    before = 2,
                                    new_col_name = "sum_binary")},
  old = {old_epi_slide(x, f = function(x, g, t) sum(x$binary),
                       before = 2,
                       new_col_name = "sum_binary")},
  iterations = 10)
# A tibble: 2 × 13
#   expression                min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#   <bch:expr>           <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>
# 1 ref_value_calculated   40.8ms   41.9ms      22.8    3.46MB     9.78     7     3
# 2 old                    20.5ms   21.1ms      46.6    2.46MB     5.17     9     1
```

For this computation, the new code is ~2x slower. With larger data sizes, the difference in runtime gets smaller (new version ~30% slower with 50k rows).

**Review by brookslogan — CHANGES_REQUESTED** [review]
Looks pretty good, and I like the more bite-size test data with more extensive tests!  Major things that need action I think are:
- ungrouped case might have a bug
- merge with dev will probably break
- same tidyeval edge cases as https://github.com/cmu-delphi/epiprocess/pull/317

File: R/slide.R [hunk 1]
```diff
@@ -223,6 +221,35 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
...
+  if (length(group_vars(x)) != 0) {
+    before_time_values_df = dplyr::cross_join(
+      # Get unique combinations of grouping columns seen in real data.
+      unique(x[, group_vars(x)]),
+      before_time_values_df
+    )
+  }
+  # Automatically fill in all other columns from `x` with `NA`s, and carry
+  # attributes over to new df.
+  before_time_values_df <- bind_rows(x[0,], before_time_values_df)
```
Comment (brookslogan) [inline R/slide.R:245]: Wow, this looks like it was pretty tricky to come up with & get the proper behavior for everything.

File: R/slide.R [hunk 2]
```diff
@@ -158,11 +159,8 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
 
   # Check that `f` takes enough args
   if (!missing(f) && is.function(f)) {
-    check_sufficient_f_args(f)
+    check_sufficient_f_args(f, 3L)
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/slide.R:162]: I need to rebase this. It's out of sync with `dev`.

File: tests/testthat/test-epi_slide.R [hunk 3]
```diff
@@ -88,13 +96,164 @@ test_that("these doesn't produce an error; the error appears only if the ref tim
...
+    filter(geo_value == "ak") %>%
+    epi_slide(before = 50,
+              slide_value = sum(.x$value))
+  expect_identical(result1, expected_output)
+})
+
+test_that("epi_slide computation via formula can use ref_time_value", {
+  expected_output = dplyr::bind_rows(
+    dplyr::tibble(geo_value = "ak", time_value = d + 1:5, value = 11:15, slide_value=as.double(d + 1:5)),
+    dplyr::tibble(geo_value = "al", time_value = d + 1:5, value = -(1:5), slide_value=as.double(d + 1:5))
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply tests/testthat/test-epi_slide.R:178]: Yep, that fixes the format of output `Date` values!

File: R/slide.R [hunk 4]
```diff
@@ -318,9 +345,15 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
 
   # If f is not missing, then just go ahead, slide by group
   if (!missing(f)) {
+    if (rlang::is_formula(f)) f = as_slide_computation(f)
+    f_rtv_wrapper = function(x, g, ...) {
+      ref_time_value = min(x$time_value) + before
+      x <- x[x$.real,]
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/slide.R:351]: Sure. Doing this is ~20% slower (median 45 vs 56 ms on a 5000-obs phony dataset).

File: R/slide.R [hunk 5]
```diff
@@ -358,5 +396,9 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
   if (!as_list_col) {
     x = unnest(x, !!new_col, names_sep = names_sep)
   }
+
+  # Drop helper column `.real`.
+  x$.real <- NULL
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/slide.R:401]: Yes! This has been a handy trick to easily improve performance. Good point last week about this being faster due to making a shallow copy vs `dplyr`'s probable deep copy.

File: R/slide.R [hunk 6]
```diff
@@ -223,6 +221,35 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
...
+  # Do set up to let us recover `ref_time_value`s later.
+  # A helper column marking real observations.
+  x$.real = TRUE
+
+  # Create df containing phony data. Df has the same columns and attributes as
+  # `x`, but filled with `NA`s aside from grouping columns. Number of rows is
+  # equal to the number of `inrange_time_values_not_in_x` we have * the
+  # number of unique levels seen in the grouping columns.
+  before_time_values_df = data.frame(time_value=inrange_time_values_not_in_x)
+  if (length(group_vars(x)) != 0) {
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/slide.R:244]: We have the dataframe `before_time_values_df` containing one column, `time_value`, of `n` dates.

If `x` is grouped and thus has `group_vars`, we need to duplicate the `n` dates in `before_time_values_df` for each unique combination of `group_vars`. This is what the `cross_join` is doing. This makes each of the phony obs and associated date in `before_time_values_df` available during [the `group_modify`](https://github.com/cmu-delphi/epiprocess/blob/0b48ea4a4ff18d9da2cade6470b424b1275be06e/R/slide.R#L364), since `group_modify` is splitting by group (AKA unique `group_vars` combinations).

If `x` is not grouped, we don't need to duplicate the `n` dates in `before_time_values_df`, since all phony obs and associated dates will be available during [the `group_modify`](https://github.com/cmu-delphi/epiprocess/blob/0b48ea4a4ff18d9da2cade6470b424b1275be06e/R/slide.R#L364). That is, `group_modify` isn't splitting the data into subsets.

I can get the `cross_join` to work more generally (currently doesn't work if `x` isn't grouped + a bit fiddly), but I was trying to avoid doing unnecessary work in the ungrouped case. If there are no `group_vars`, we'd end up joining `before_time_values_df` onto an empty df and expecting back the original version of `before_time_values_df`.

> what happens in the else case? I would have thought that... we need special processing

Were you thinking of something in particular?

File: R/slide.R [hunk 7]
```diff
@@ -223,6 +221,35 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
     after <- time_step(after)
   }
 
+  inrange_time_values_not_in_x = ref_time_values - before
```
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/slide.R:232]: Rename to
```
  min_ref_time_values = ref_time_values - before
  min_ref_time_values_not_in_x <- min_ref_time_values[!(min_ref_time_values %in% unique(x$time_value))]
```
?

**Comment by nmdefries** [comment]
I won't be able to finish the last few issues here before my PTO. Will pick those up when I get back.

File: R/slide.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (brookslogan) [reply R/slide.R]: `min_ref` might be confusing, since these aren't actually from ref time values.  Trying `start` for now.

File: R/slide.R [hunk 8]
```diff
@@ -223,6 +221,35 @@ epi_slide = function(x, f, ..., before, after, ref_time_values,
...
+
+  # Create df containing phony data. Df has the same columns and attributes as
+  # `x`, but filled with `NA`s aside from grouping columns. Number of rows is
+  # equal to the number of `inrange_time_values_not_in_x` we have * the
+  # number of unique levels seen in the grouping columns.
+  before_time_values_df = data.frame(time_value=inrange_time_values_not_in_x)
+  if (length(group_vars(x)) != 0) {
+    before_time_values_df = dplyr::cross_join(
+      # Get unique combinations of grouping columns seen in real data.
+      unique(x[, group_vars(x)]),
```
[the remark this answers is not in the document]
Comment (brookslogan) [reply R/slide.R:247]: Tried to test this here:
```{r}
as_epi_df(tibble(geo_value=c("a","a","a","b"), time_value = c(1:3, 2))) %>%
  group_by(geo_value) %>%
  epi_slide(function(x,g,t) if(nrow(x) == 0L) stop("0 row") else 1, before = 0, after = 0)
```
and got no error, so maybe this isn't happening.  It looks like the first part of `slide_one_grp` is already taking care of this.

File: R/slide.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (brookslogan) [reply R/slide.R:244]: I might have been thinking of duplication regarding non-grouping key variables? But that's not handled in the grouped case, and not necessary in either case.

However, I am surprised how we seem to get the right recycling behavior in either case in this type of situation.  I don't see a `.real` filter being applied when determining the recyling counts.  Trying to figure out why it works.

[the remark this answers is not in the document]
Comment (brookslogan) [reply R/slide.R:244]: Ah, I see, it's also due to the first part of `slide_one_grp` which mutates the `time_values` binding.

**Comment by brookslogan** [comment]
An interesting performance experiment...  I thought we might speed up things by changing `.x[.x$.real,]` to `.x[.x[[".real"]],]` as I thought it's supposed to be faster based on some R6 benchmarks that include comparisons of these things, but it's actually slower and doubles the amount of garbage collections!
```{r}
bench::mark(
  grped %>% epi_slide(cases_7dav = mean(cases), before=6),
  grped %>% curr_epi_slide(cases_7dav = mean(cases), before=6)
)
#> # A tibble: 2 × 13
#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#>   <bch:expr>    <bch> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
#> 1 grped %>% ep… 3.13s  3.13s     0.319      26MB     6.38     1    20      3.13s
#> 2 grped %>% cu… 2.96s  2.96s     0.338      26MB     3.04     1     9      2.96s
#> # ℹ 4 more variables: result <list>, memory <list>, time <list>, gc <list>
#> Warning message:
#> Some expressions had a GC in every iteration; so filtering is disabled. 
```

**Review by brookslogan — APPROVED** [review]
Thanks for clearing up some of my confusion.  I've applied some minor updates and am going to merge.  Docs and NEWS might still need a little work; I'll try to tackle that in after the merge.

File: R/slide.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (nmdefries) [reply R/slide.R:250]: Great question about potential 0-row inputs! I don't see a way around the `cross_join`, but maybe we could add a short-circuit condition inside the [`f` wrapper functions](https://github.com/cmu-delphi/epiprocess/blob/0b48ea4a4ff18d9da2cade6470b424b1275be06e/R/slide.R#L357-L362). Like
```
if (all(!x$.real)) { return() }
```

It could be worth adding a test for this, too. If we change upstream behavior, we don't want to start processing 0-row inputs when we weren't before.
